### PR TITLE
Inline docs

### PR DESCRIPTION
**Pull request by simonbyrne**

### CODE REVIEW

File: src/glmfit.jl [hunk 1]
```diff
@@ -1,5 +1,5 @@
 type GlmResp{V<:FPVector,D<:UnivariateDistribution,L<:Link} <: ModResp       # response in a glm model
-    y::V                                       # response
+   y::V                                       # response
```
Comment (simonbyrne) [inline src/glmfit.jl]: This somehow got unaligned

**Comment by simonbyrne** [comment]
@abhijithch Looks good. See my comment, and it needs to be rebased (also you can squash the commits).

**Comment by KristofferC** [comment]
Maybe you already know about this but the one who merges can squash the commits on merge with the github UI now.

**Comment by simonbyrne** [comment]
Ah that's, good to know.

File: src/glmfit.jl [hunk 2]
```diff
@@ -31,6 +31,15 @@ end
 # returns the sum of the squared deviance residuals
 deviance(r::GlmResp) = sum(r.devresid)
 
+"""
+    updatemu!{T<:FPVector}(r::GlmResp{T}, linPr::T)
+
+Update the response type from the linear predictor.
+
+#### Arguments:
```
Comment (nalimilan) [inline src/glmfit.jl]: I the recent [docstring style guide](http://docs.julialang.org/en/latest/manual/documentation/), we agreed on the principle that an `Arguments` section should only be used when there are so many of them that presenting them in the description isn't possible.

Here, I think it would be better to document them in the description, which is much shorter:
```
Update the response type `r` from the linear predictor `linPr`.
```


Also, isn't this the "GLM response" rather than the "GLM response type"?

File: src/glmfit.jl [hunk 3]
```diff
@@ -67,6 +76,14 @@ function updatemu!{T<:FPVector}(r::GlmResp{T}, linPr::T)
     r
 end
 
+"""
+    wrkresp(r::GlmResp)
+
+Working response.
```
Comment (nalimilan) [inline src/glmfit.jl]: Again, the style guide says the description should start with an imperative verb. So:
```
Return the working response for GLM response `r`.
```

File: src/glmfit.jl [hunk 4]
```diff
@@ -230,6 +247,41 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
                     d::UnivariateDistribution, l::Link=canonicallink(d); kwargs...) =
     fit(M, float(X), float(y), d, l; kwargs...)
 
+"""
+    glm(X, y, args...; kwargs...)
+
+Fits the generalised linear model, specified by the symbolic description of the
```
Comment (nalimilan) [inline src/glmfit.jl]: American English "generalized" is used everywhere else.

X isn't a symbolic description here AFAIK, it's the model matrix/design matrix.

File: src/glmfit.jl [hunk 5]
```diff
@@ -230,6 +247,41 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
+model along with the family of distribution and a link function.
+
+#### Arguments:
+* `X`: The formula which is the symbolic representation of the model to fit.
+Uses column symbols from the DataFrame data, for example, if names(data)
+=[:Y,:X1,:X2], then a valid formula is Y~X1+X2
+* `y`: A DataFrame which may contain NA values, any rows with NA values are ignored.
+* `family`: Specifies the choice of variance, can be `Bernoulli()`, `Binomial()`,
```
Comment (nalimilan) [inline src/glmfit.jl]: `family` and `link` should be added to the signature, even if they are technically passed to another function. Please also document `dofit`, `wts` and `offset`.

File: src/glmfit.jl [hunk 6]
```diff
@@ -230,6 +247,41 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
+    glm(X, y, args...; kwargs...)
+
+Fits the generalised linear model, specified by the symbolic description of the
+model along with the family of distribution and a link function.
+
+#### Arguments:
+* `X`: The formula which is the symbolic representation of the model to fit.
+Uses column symbols from the DataFrame data, for example, if names(data)
+=[:Y,:X1,:X2], then a valid formula is Y~X1+X2
+* `y`: A DataFrame which may contain NA values, any rows with NA values are ignored.
```
Comment (nalimilan) [inline src/glmfit.jl]: `y` is the response vector.

File: src/glmfit.jl [hunk 7]
```diff
@@ -230,6 +247,41 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
+* `family`: Specifies the choice of variance, can be `Bernoulli()`, `Binomial()`,
+`Gamma()`, `Normal()` or `Poisson()` 
+* `link`: This function provides the link function for example, LogitLink() is a valid link for the Binomial() family.
```
Comment (nalimilan) [inline src/glmfit.jl]: Instead of examples, please provide the full like, just like for distributions.

File: src/glmfit.jl [hunk 8]
```diff
@@ -230,6 +247,41 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
+
+#### Examples:
+```
+julia> data=DataFrame(X=[1,2,3],Y=[2,4,7])
```
Comment (nalimilan) [inline src/glmfit.jl]: With the latest DataFrames.jl, this looks like this:
```julia
3×2 DataFrames.DataFrame
│ Row │ X │ Y │
┝━━━━━┿━━━┿━━━┥
│ 1   │ 1 │ 2 │
│ 2   │ 2 │ 4 │
│ 3   │ 3 │ 7 │
```

File: src/glmtools.jl [hunk 9]
```diff
@@ -11,6 +11,27 @@ type LogLink <: Link end
 type ProbitLink <: Link end
 type SqrtLink <: Link end
 
+"""
+    linkfun(::Link, μ)
+
+Link function mapping μ to η (the linear predictor).
```
Comment (nalimilan) [inline src/glmtools.jl]: `Return the Link function...`. And same below.

File: src/glmtools.jl [hunk 10]
```diff
@@ -11,6 +11,27 @@ type LogLink <: Link end
...
+    linkfun(::Link, μ)
+
+Link function mapping μ to η (the linear predictor).    
+"""
+function linkfun end
+
+"""
+    linkinv(::Link, η)
+
+Inverse Link function mapping mu to eta, the linear predictor (CauchitLink response model).
```
Comment (nalimilan) [inline src/glmtools.jl]: For consistency with above, `mapping μ to η` (the linear predictor). AFAIK `CauchitLink|` is only one possible link, so just remove this part.

File: src/glmtools.jl (hunk 9 above)
Comment (nalimilan) [inline src/glmtools.jl]: Also, `::Link` is incorrect. You could write it as `Link`, and change the description to something like:
```julia
Compute η (the linear predictor) from μ using link function of type `Link`.
```

The same pattern should be used for similar functions below.

File: src/glmtools.jl [hunk 11]
```diff
@@ -43,6 +64,16 @@ linkfun(::SqrtLink, μ) = sqrt(μ)
 linkinv(::SqrtLink, η) = abs2(η)
 mueta(::SqrtLink, η) = 2η
 
+"""
+    canonicallink(d)
+
+Canonical link function for a `d` distribution.
```
Comment (nalimilan) [inline src/glmtools.jl]:
```
Get the canonical link function for distribution `d`. Supported distributions include `Bernoulli()`, `Binomial()`, `Gamma()`, `Normal()` and `Poisson()`.
```

File: src/linpred.jl [hunk 12]
```diff
@@ -1,4 +1,9 @@
 ## Return the linear predictor vector
+"""
+    linpred!(out, p::LinPred, f::Real=1.)
+
+Linear predictor inplace.
```
Comment (nalimilan) [inline src/linpred.jl]: Here and below, use the imperative form `Compute...`.

Comment (nalimilan) [inline src/linpred.jl]: And explain what the arguments mean.

File: src/linpred.jl [hunk 13]
```diff
@@ -113,14 +128,44 @@ end
...
+"""
+    coef(x)
+
+Extract the estimates of the coefficients in the model.
+
+#### Arguments:
+* `x`: Inputs can either `LinPred` or `LinPredModel`.
+
+#### Examples:
+Refer example for `glm()`
```
Comment (nalimilan) [inline src/linpred.jl]: A partial example isn't terribly practical. Probably remove it completely, as this is such basic functionality that it is covered elsewhere.

File: src/linpred.jl [hunk 14]
```diff
@@ -113,14 +128,44 @@ end
...
+The output is an array, where the first element is the intercept 
+and the second element is the coefficient of the `x` component.
+"""
+function coef end
+
 coef(x::LinPred) = x.beta0
 coef(x::LinPredModel) = coef(x.pp)
 
+"""
+Degrees of freedom for residuals, when meaningful.
```
Comment (nalimilan) [inline src/linpred.jl]: No, these are the "residual degrees of freedom". "When meaningful" should be removed (could be added to any Julia function).

File: src/linpred.jl [hunk 15]
```diff
@@ -113,14 +128,44 @@ end
...
+function df_residual end
+
 df_residual(x::LinPredModel) = df_residual(x.pp)
 df_residual(x::@compat(Union{DensePred,SparsePredChol})) = size(x.X, 1) - length(x.beta0)
 
 invchol(x::DensePred) = inv(cholfact!(x))
 invchol(x::SparsePredChol) = cholfact!(x)\eye(size(x.X, 2))
+
+"""
+Estimated variance-covariance matrix of the coefficient estimates.
```
Comment (nalimilan) [inline src/linpred.jl]: Missing signature and imperative form.

File: src/lm.jl [hunk 16]
```diff
@@ -12,13 +12,23 @@ type LmResp{V<:FPVector} <: ModResp  # response in a linear model
 end
 convert{V<:FPVector}(::Type{LmResp{V}}, y::V) = LmResp{V}(fill!(similar(y), zero(eltype(V))), similar(y, 0), similar(y, 0), y)
 
+"""
+    updatemu!{V<:FPVector}(r::LmResp{V}, linPr::V)
+
+Update the response `r` from the linear predictor, `V`.
```
Comment (nalimilan) [inline src/lm.jl]: No need for a comma here.

File: src/lm.jl [hunk 17]
```diff
@@ -12,13 +12,23 @@ type LmResp{V<:FPVector} <: ModResp  # response in a linear model
...
     n = length(linPr); length(r.y) == n || error("length(linPr) is $n, should be $(length(r.y))")
     length(r.offset) == 0 ? copy!(r.mu, linPr) : broadcast!(+, r.mu, linPr, r.offset)
     deviance(r)
 end
 updatemu!{V<:FPVector}(r::LmResp{V}, linPr) = updatemu!(r, convert(V,vec(linPr)))
 
+"""
+    deviance(r::LmResp)
+
+Measure of the model fit, weighted residual sum of squares for lm's
```
Comment (nalimilan) [inline src/lm.jl]: Already documented in StatsBase, just skip it.

File: src/lm.jl [hunk 18]
```diff
@@ -71,7 +81,15 @@ function fit(::Type{LinearModel}, X::Matrix, y::Vector)
     return fit(LinearModel{LmResp{typeof(yy)}, DensePredQR{T}}, X, y)
 end
 
+"""
+    lm()
+Fit a linear model using QR factorization.
```
Comment (nalimilan) [inline src/lm.jl]: Missing signature, description or arguments, and line break.

File: src/lm.jl [hunk 19]
```diff
@@ -71,7 +81,15 @@ function fit(::Type{LinearModel}, X::Matrix, y::Vector)
...
+"""
 lm(X, y) = fit(LinearModel, X, y)
+"""
+    lmc()
+Fit a linear model using Cholesky factorization.
```
Comment (nalimilan) [inline src/lm.jl]: Same here.

File: src/lm.jl [hunk 20]
```diff
@@ -90,6 +108,15 @@ function coeftable(mm::LinearModel)
               ["x$i" for i = 1:size(mm.pp.X, 2)], 4)
 end
 
+"""
+    predict(mm, newx)
+
+Returns the predictions from the fitted model for a new set of independent variables.
```
Comment (nalimilan) [inline src/lm.jl]: ``
Return the predictions from a fitted model `mm` based on a matrix of independent variables `newx`. Each row of `newx` represents an observation for which to predict the response, and each column represents an independent variable.
``

Just remove the `Arguments` section.

File: test/runtests.jl [hunk 21]
```diff
@@ -17,6 +17,24 @@ test_show(lm1)
 @test_approx_eq vcov(lm1) Σ
 @test_approx_eq cor(lm1.model) diagm(diag(Σ))^(-1/2)*Σ*diagm(diag(Σ))^(-1/2)
 
+#1).
+data=DataFrame(X=[1,2,3],Y=[2,4,7])
```
Comment (nalimilan) [inline test/runtests.jl]: On master the tests include many other values like fit statistics. I'd say better leave this for another PR.

**Comment by nalimilan** [comment]
Thanks for the update, this will be very helpful!

File: src/glmfit.jl (hunk 7 above)
Comment (abhijithch) [inline src/glmfit.jl]: Do you mean a link to the example?

Comment (nalimilan) [inline src/glmfit.jl]: Oops, I meant "the full *list*" of types.

File: src/glmfit.jl (hunk 2 above)
Comment (simonbyrne) [inline src/glmfit.jl]: "GLM response object" would probably be better.

File: src/linpred.jl (hunk 12 above)
Comment (abhijithch) [inline src/linpred.jl]: Hi I am not very sure on `f::Real=1`. Could you explain this in brief.

Comment (nalimilan) [inline src/linpred.jl]: I must confess I don't know. @dmbates, could you share your expertise?

Comment (dmbates) [inline src/linpred.jl]: The step factor `f` determines the parameter value at which the linear predictor is evaluated in an iterative algorithm like IRLS.  The current coefficient vector is `beta0` and the proposed increment is `delbeta`.  The linear predictor is evaluated at `beta0 + f .* delbeta`

**Comment by abhijithch** [comment]
@nalimilan could you please review and merge?

**Comment by simonbyrne** [comment]
This still needs to be rebased against master.

File: src/glmtools.jl [hunk 22]
```diff
@@ -58,11 +87,34 @@ glmvar(::Gamma, ::Link, μ, η) = abs2(μ)
 glmvar(::Normal, ::Link, μ, η) = 1
 glmvar(::Poisson, ::Link, μ, η) = μ
 
+"""
+    mustart(d, y, wt)
+
+Derive starting values for the mu vector.
+
+#### Arguments:                                                                             
+* `d`: Distribution, can be any one of `::Binomial`, `::Gamma`, `::Normal` or `::Poisson`  * `y`:                                                                                     * `wt`:
```
Comment (simonbyrne) [inline src/glmtools.jl]: y is the response vector
wt is the input weight vector

File: src/glmtools.jl [hunk 23]
```diff
@@ -58,11 +87,34 @@ glmvar(::Gamma, ::Link, μ, η) = abs2(μ)
...
+    devresid(d, y, μ, wt)
+
+Return a vector of squared deviance residuals.
+
+#### Arguments:
+* `d`: Distribution, can be any one of `::Binomial`, `::Gamma`, `::Normal` or `::Poisson`
+* `y`: 
+* `μ`:
+* `wt`:
+"""
```
Comment (simonbyrne) [inline src/glmtools.jl]: y is the response vector
mu is the predicted mean vector
wt is the weight vector

Comment (nalimilan) [inline src/glmtools.jl]: More precisely, `wt` provides frequency weights a.k.a. case weights.

File: src/glmfit.jl [hunk 24]
```diff
@@ -230,6 +240,43 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
+
+#### Arguments:
+* `X::Matrix`: The formula which is the symbolic representation of the model to fit.
+Uses column symbols from the DataFrame data, for example, if names(data)
+=[:Y,:X1,:X2], then a valid formula is Y~X1+X2.
+* `y::AbstractVector`: Response vector.
+* `family::UnivariateDistribution`: Specifies the choice of variance, can be `Bernoulli()`, `Binomial()`, `Gamma()`, `Normal()` or `Poisson()`.
+* `link::Link=canonicallink(d)`: This function provides the link function for example, LogitLink() is a valid link for the Binomial() family.
+* `dofit::Bool=true`: Indicates whether to fit the model or not.
+* `wts::V=fill!(similar(y), one(eltype(y)))`: Weights (inverse dispersion).
```
Comment (nalimilan) [inline src/glmfit.jl]: AFAIK these are frequency weights a.k.a. case weights, not inverse dispersion weights.

File: src/glmfit.jl [hunk 25]
```diff
@@ -230,6 +240,43 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
     fit(M, float(X), float(y), d, l; kwargs...)
 
+"""
+    glm(X, y, family, link; wts, offset, dofit)
+
+Fits the generalized linear model, specified by the symbolic description of the
+model along with the family of distribution and a link function.
+
+#### Arguments:
+* `X::Matrix`: The formula which is the symbolic representation of the model to fit.
```
Comment (nalimilan) [inline src/glmfit.jl]: This is still incorrect (see my previous comments).

File: src/glmfit.jl [hunk 26]
```diff
@@ -230,6 +240,43 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
+Uses column symbols from the DataFrame data, for example, if names(data)
+=[:Y,:X1,:X2], then a valid formula is Y~X1+X2.
+* `y::AbstractVector`: Response vector.
+* `family::UnivariateDistribution`: Specifies the choice of variance, can be `Bernoulli()`, `Binomial()`, `Gamma()`, `Normal()` or `Poisson()`.
```
Comment (nalimilan) [inline src/glmfit.jl]: Specifies the choice of variance -> Probability distribution family

File: src/glmfit.jl [hunk 27]
```diff
@@ -230,6 +240,43 @@ fit{M<:AbstractGLM}(::Type{M}, X::@compat(Union{Matrix,SparseMatrixCSC}), y::Abs
...
+* `link::Link=canonicallink(d)`: This function provides the link function for example, LogitLink() is a valid link for the Binomial() family.
```
Comment (nalimilan) [inline src/glmfit.jl]: "This function provides" isn't useful here. Also, instead of an example, please list all supported links.

Finally, lines must be wrapped at 92 chars.

File: src/glmtools.jl [hunk 28]
```diff
@@ -11,6 +11,27 @@ type LogLink <: Link end
...
+    linkfun(Link, μ)
+
+Compute η (the linear predictor) from μ using link function of type `Link`.
+"""
+function linkfun end
+
+"""
+    linkinv(Link, η)
+
+Return the inverse Link function mapping μ(mean predictor) to η(the linear predictor).
```
Comment (nalimilan) [inline src/glmtools.jl]: Space before `(`. Also use "the" consistently.

File: src/glmtools.jl (hunk 22 above)
Comment (nalimilan) [inline src/glmtools.jl]: Same remark as above regarding weights. `d` must be e.g. `Binomial()`, not `::Binomial`. Also list ` Bernouilli()` after rebasing.

File: src/glmtools.jl [hunk 29]
```diff
@@ -58,11 +87,34 @@ glmvar(::Gamma, ::Link, μ, η) = abs2(μ)
...
 mustart(::Normal, y, wt) = y
 mustart(::Poisson, y, wt) = y + oftype(y, 0.1)
 
+"""
+    devresid(d, y, μ, wt)
+
+Return a vector of squared deviance residuals.
+
+#### Arguments:
+* `d`: Distribution, can be any one of `::Binomial`, `::Gamma`, `::Normal` or `::Poisson`
```
Comment (nalimilan) [inline src/glmtools.jl]: Same here.

File: src/linpred.jl [hunk 30]
```diff
@@ -113,14 +132,34 @@ end
 Base.cholfact{T}(p::SparsePredChol{T}) = copy(p.chol)
 Base.cholfact!{T}(p::SparsePredChol{T}) = p.chol
 
+"""
+    coef(x)
```
Comment (nalimilan) [inline src/linpred.jl]: This function is defined in StatsBase and should be documented there.

File: src/linpred.jl [hunk 31]
```diff
@@ -113,14 +132,34 @@ end
...
+
+Extract the estimates of the coefficients in the model `x` which is either
+`LinPred` or `LinPredModel`.
+"""
+function coef end
+
 coef(x::LinPred) = x.beta0
 coef(x::LinPredModel) = coef(x.pp)
 
+"""
```
Comment (nalimilan) [inline src/linpred.jl]: Same for this one.

File: src/linpred.jl [hunk 32]
```diff
@@ -113,14 +132,34 @@ end
...
+function df_residual end
+
 df_residual(x::LinPredModel) = df_residual(x.pp)
 df_residual(x::@compat(Union{DensePred,SparsePredChol})) = size(x.X, 1) - length(x.beta0)
 
 invchol(x::DensePred) = inv(cholfact!(x))
 invchol(x::SparsePredChol) = cholfact!(x)\eye(size(x.X, 2))
+
+"""
```
Comment (nalimilan) [inline src/linpred.jl]: Same for this one.

File: src/linpred.jl [hunk 33]
```diff
@@ -134,6 +173,11 @@ function cor(x::LinPredModel)
     scale!(invstd, scale!(Σ, invstd))
 end
 
+"""
```
Comment (nalimilan) [inline src/linpred.jl]: ...and this one, as well as `formula` and `nobs`.

File: src/lm.jl [hunk 34]
```diff
@@ -12,6 +12,11 @@ type LmResp{V<:FPVector} <: ModResp  # response in a linear model
 end
 convert{V<:FPVector}(::Type{LmResp{V}}, y::V) = LmResp{V}(fill!(similar(y), zero(eltype(V))), similar(y, 0), similar(y, 0), y)
 
+"""
+    updatemu!{V<:FPVector}(r::LmResp{V}, linPr::V)
+
+Update the response `r` from the linear predictor `V`.
```
Comment (nalimilan) [inline src/lm.jl]: V -> linPr

File: src/lm.jl [hunk 35]
```diff
@@ -71,7 +76,18 @@ function fit(::Type{LinearModel}, X::Matrix, y::Vector)
     return fit(LinearModel{LmResp{typeof(yy)}, DensePredQR{T}}, X, y)
 end
 
+"""
+    lm(X::Matrix, y::Vector)
+
+Fit a linear model using QR factorization.
```
Comment (nalimilan) [inline src/lm.jl]: Missing description of arguments.

**Comment by andreasnoack** [comment]
Bump

**Comment by andreasnoack** [comment]
Bump

@simonbyrne

**Comment by nalimilan** [comment]
Re-bump ! :-)

**Comment by abhijithch** [comment]
Missed this one, will look into this soon.

**Comment by dmbates** [comment]
Where are we on this PR?  I am willing to work on it if someone can point me in the direction of what needs to be done.

**Comment by nalimilan** [comment]
@dmbates That would be great. I think the PR "just" needs to be rebased and to take into account review comments. The advantage is that you'll know the meaning of functions better than those who reviewed it!

**Comment by dmbates** [comment]
Sorry for asking such an elementary question but how would I go about getting the branch with the modifications?  Do I need to clone the fork the repository from @abhijithch?

**Comment by ararslan** [comment]
```
git remote add abhijithch https://github.com/abhijithch/GLM.jl.git
git checkout abhijithch/inlinedoc
git checkout -b inlinedoc
```
Now you have a local branch `inlinedoc` that tracks the corresponding branch on Abhijith's fork. You can make changes, then `git push abhijith/inlinedoc`, which you have write access to as someone with commit access to this repository.

**Comment by andreasnoack** [comment]
Superseded by #189
